src/main/scala/cognite/spark/v1/udf/CogniteUdfs.scala

```diff
@@ -0,0 +1,143 @@
+package cognite.spark.v1.udf
+
+import cats.effect.IO
+import cognite.spark.v1.Constants.{DefaultBaseUrl, DefaultMaxRetries}
+import cognite.spark.v1.{CdfSparkException, CdpConnector, Constants}
+import com.cognite.sdk.scala.common.ApiKeyAuth
+import com.cognite.sdk.scala.v1.{FunctionCall, GenericClient}
+import com.softwaremill.sttp.SttpBackend
+import io.circe.{Json, JsonObject, parser}
+import org.apache.spark.sql.SparkSession
+
+class CogniteUdfs(sparkSession: SparkSession) extends Serializable {
+
+  def initializeUdfs(
+      apiKey: ApiKeyAuth,
+      baseUrl: Option[String] = None,
+      maxRetries: Option[Int] = None): Unit = {
+    sparkSession.udf.register(
+      "cdf_function",
+      (functionId: Long, data: String) =>
+        callFunctionAndGetResult(
+          functionId,
+          data,
+          apiKey,
+          baseUrl.getOrElse(DefaultBaseUrl),
+          maxRetries.getOrElse(DefaultMaxRetries)
+        ).unsafeRunSync()
+    )
+    sparkSession.udf.register(
+      "cdf_function_external_id",
+      (externalId: String, data: String) =>
+        callFunctionByExternalId(
+          externalId,
+          data,
+          apiKey,
+          baseUrl.getOrElse(DefaultBaseUrl),
+          maxRetries.getOrElse(DefaultMaxRetries)
+        ).unsafeRunSync()
+    )
+    sparkSession.udf.register(
+      "cdf_function_name",
+      (name: String, data: String) =>
+        callFunctionByName(
+          name,
+          data,
+          apiKey,
+          baseUrl.getOrElse(DefaultBaseUrl),
+          maxRetries.getOrElse(DefaultMaxRetries)
+        ).unsafeRunSync()
+    )
+  }
+
+  private def getFunctionResult(
+      client: GenericClient[IO],
+      functionId: Long,
+      result: FunctionCall): IO[Json] = {
+    var res = result
+    var i = 0
+    while (res.status.isDefined && res.status.get == "Running" && i < 30) {
+      i += 1
+      res = client
+        .functionCalls(functionId)
+        .retrieveById(res.id.get)
+        .unsafeRunSync()
+    }
+
+    if (res.status.isDefined && res.status.get == "Completed") {
+      for {
+        functionCallResponse <- client
+          .functionCalls(functionId)
+          .retrieveResponse(res.id.getOrElse(0L))
+      } yield functionCallResponse.response.getOrElse(Json.fromJsonObject(JsonObject.empty))
+    } else {
+      val logs = client.functionCalls(functionId).retrieveLogs(res.id.get).unsafeRunSync()
+      throw new CdfSparkException(
+        "CDF function call failed. Res = " + res.toString + "logs = " + logs.items.toString())
```

Comment:
Missing a space in `"logs = "`, should be (adding a comma, as well) `", logs = "`

Reply:
I added logs just to debug something, and then I removed it now. Is it something we should include?